### ISSUE TITLE
Unusable on foreign keyboard

### ISSUE THREAD

**Issue by leancode** [rbreaves/kinto#192]
**Describe the bug**
I am a switcher from MacOS to Ubuntu and found Kinto. All exited I installed Kinto and unfortunately it promptly made my computer unusable.
How?
It removed my ability to type the pipe |, the @, #, ~ and backslash \ symbols to name the most important ones.
As you can imagine those symbols are very important to use the system.
I turned on the on-screen keyboard to be able to continue using my system. 

**Expected behavior**
The right alt key should continue to operate as it does when I am using my mac: Be able to continue to use the 3rd row keys.

**Install Type:** Bare Metal

**Distro:** Ubuntu 20.04 LTS

**DE:** Default Gnome

**Keyboard:** Swiss

**Branch/Commit:**
commit 8dc05e7acecd8ff7e57a585f2159bc5767b41251 (HEAD -> master, origin/master, origin/HEAD)

**SOLUTION:**
In ~/.config/kinto/kinto.py
change this:
```
# [Global modemap] Change modifier keys as in xmodmap
define_conditional_modmap(lambda wm_class: wm_class.casefold() not in terminals,{
    # # Chromebook
    # Key.LEFT_ALT: Key.RIGHT_CTRL,   # Chromebook
    # Key.LEFT_CTRL: Key.LEFT_ALT,    # Chromebook
    # Key.RIGHT_ALT: Key.RIGHT_CTRL,  # Chromebook
    # Key.RIGHT_CTRL: Key.RIGHT_ALT,  # Chromebook

    # # Default Mac/Win
    Key.LEFT_ALT: Key.RIGHT_CTRL,   # WinMac
    Key.LEFT_META: Key.LEFT_ALT,    # WinMac
    Key.LEFT_CTRL: Key.LEFT_META,   # WinMac
    Key.RIGHT_ALT: Key.RIGHT_CTRL,  # WinMac
    Key.RIGHT_META: Key.RIGHT_ALT,  # WinMac
    Key.RIGHT_CTRL: Key.RIGHT_META, # WinMac

    # # KintoWin

    # # Mac Only
    # Key.LEFT_META: Key.RIGHT_CTRL,  # Mac
    # Key.LEFT_CTRL: Key.LEFT_META,   # Mac
    # Key.RIGHT_META: Key.RIGHT_CTRL, # Mac
    # Key.RIGHT_CTRL: Key.RIGHT_META, # Mac
})
```
to
```
define_conditional_modmap(lambda wm_class: wm_class.casefold() not in terminals,{
    # # Chromebook
    # Key.LEFT_ALT: Key.RIGHT_CTRL,   # Chromebook
    # Key.LEFT_CTRL: Key.LEFT_ALT,    # Chromebook
    # Key.RIGHT_ALT: Key.RIGHT_CTRL,  # Chromebook
    # Key.RIGHT_CTRL: Key.RIGHT_ALT,  # Chromebook

    # # Default Mac/Win
    Key.LEFT_ALT: Key.RIGHT_CTRL,   # WinMac
    Key.LEFT_META: Key.LEFT_ALT,    # WinMac
    Key.LEFT_CTRL: Key.LEFT_META,   # WinMac
#   Key.RIGHT_ALT: Key.RIGHT_CTRL,  # WinMac
    Key.RIGHT_META: Key.RIGHT_ALT,  # WinMac
    Key.RIGHT_CTRL: Key.RIGHT_META, # WinMac

    # # KintoWin

    # # Mac Only
    # Key.LEFT_META: Key.RIGHT_CTRL,  # Mac
    # Key.LEFT_CTRL: Key.LEFT_META,   # Mac
    # Key.RIGHT_META: Key.RIGHT_CTRL, # Mac
    # Key.RIGHT_CTRL: Key.RIGHT_META, # Mac
})
```
and change this:
```
# [Conditional modmap] Change modifier keys in certain applications
define_conditional_modmap(re.compile(termStr, re.IGNORECASE), {
    # # Chromebook
    # Key.LEFT_ALT: Key.RIGHT_CTRL,     # Chromebook
    # # Left Ctrl Stays Left Ctrl
    # Key.LEFT_META: Key.LEFT_ALT,      # Chromebook
    # Key.RIGHT_ALT: Key.RIGHT_CTRL,    # Chromebook
    # Key.RIGHT_CTRL: Key.RIGHT_ALT,    # Chromebook
    # # Right Meta does not exist on chromebooks

    # # Default Mac/Win
    Key.LEFT_ALT: Key.RIGHT_CTRL,   # WinMac
    Key.LEFT_META: Key.LEFT_ALT,    # WinMac
    Key.LEFT_CTRL: Key.LEFT_CTRL,   # WinMac
    Key.RIGHT_ALT: Key.RIGHT_CTRL,  # WinMac
    Key.RIGHT_META: Key.RIGHT_ALT,  # WinMac
    Key.RIGHT_CTRL: Key.LEFT_CTRL,  # WinMac

    # # KintoWin
    # Key.LEFT_CTRL: Key.RIGHT_CTRL,   # KintoWin
    # Key.LEFT_META: Key.LEFT_CTRL,    # KintoWin

    # # Mac Only
    # Key.LEFT_META: Key.RIGHT_CTRL,  # Mac
    # # Left Ctrl Stays Left Ctrl
    # Key.RIGHT_META: Key.RIGHT_CTRL, # Mac
    # Key.RIGHT_CTRL: Key.LEFT_CTRL,  # Mac
})
```
to:
```
# [Conditional modmap] Change modifier keys in certain applications
define_conditional_modmap(re.compile(termStr, re.IGNORECASE), {
    # # Chromebook
    # Key.LEFT_ALT: Key.RIGHT_CTRL,     # Chromebook
    # # Left Ctrl Stays Left Ctrl
    # Key.LEFT_META: Key.LEFT_ALT,      # Chromebook
    # Key.RIGHT_ALT: Key.RIGHT_CTRL,    # Chromebook
    # Key.RIGHT_CTRL: Key.RIGHT_ALT,    # Chromebook
    # # Right Meta does not exist on chromebooks

    # # Default Mac/Win
    Key.LEFT_ALT: Key.RIGHT_CTRL,   # WinMacit
    Key.LEFT_META: Key.LEFT_ALT,    # WinMac
    Key.LEFT_CTRL: Key.LEFT_CTRL,   # WinMac
#   Key.RIGHT_ALT: Key.RIGHT_CTRL,  # WinMac
    Key.RIGHT_META: Key.RIGHT_ALT,  # WinMac
    Key.RIGHT_CTRL: Key.LEFT_CTRL,  # WinMac

    # # KintoWin
    # Key.LEFT_CTRL: Key.RIGHT_CTRL,   # KintoWin
    # Key.LEFT_META: Key.LEFT_CTRL,    # KintoWin

    # # Mac Only
    # Key.LEFT_META: Key.RIGHT_CTRL,  # Mac
    # # Left Ctrl Stays Left Ctrl
    # Key.RIGHT_META: Key.RIGHT_CTRL, # Mac
    # Key.RIGHT_CTRL: Key.LEFT_CTRL,  # Mac
})
```
then restart the service with:
```
sudo systemctl restart xkeysnail
```
Now its working PERFECTLY. I love it. All my muscle memory from years of working on a mac is working on Ubuntu and its saving me oodles of time. Thank you very much.

I think we need just need another install option to ask about the use of the AltGr key on the right.

**Comment by rbreaves** [comment]
I will definitely implement this change, seems obvious now that you’re pointing it out 😅.

**Comment by rbreaves** [comment]
Looking at 3 keyboards I have I am not sure that a mac user would always want the right Alt key to be AltGr, the Alt is sometimes directly right of the space bar and sometimes isn't, but I will add in a prompt to ask the question and then apply the keymap assuming they want the right Alt key left alone.

**Comment by rbreaves** [comment]
A better fix will likely come through a wizard asking them to press the keys 1 by 1 and telling the program exactly how they want each key mapped modifier wise. Probably the only true way to make Kinto install perfectly every time.

I already planned to do that on the keys on the left side, which was simple as they are more standardized than the keys on the right side of the spacebar. The same method of prompting and setting should work though. For now a simple prompt has been added till I rewrite that portion of Kinto.

**Comment by rbreaves** [comment]
I know I already closed this issue, but did the commit 9 days ago fix this issue for you? (Technically I think the Right Control key did remain to be AltGr so it was never really lost - but if it feels more natural to have in the Right Alt position then that option is there now).

@leancode

**Comment by leancode** [comment]
@rbreaves Hi Ben, sorry for the delay in getting back on this. I just set up another linux machine using the latest changes you made. I noticed the question about the right alt key and it made a change in that its working in the terminal but not in the browser. A step in the right direction.

I looked at the file to see what changes the script had implemented and I noticed that the global mod map still had the alt key change not yet commented out and I did that restarted and all was well.

I did notice that the conditional mod map was commented out by the installer already.

So what would need to be done is to include the change of commenting out the global mod map for the alt right in the installer and I think the issue is truly licked.

Thanks for your responsiveness. This really is a treat to work with.

**Comment by rbreaves** [comment]
Hmmm well that was a surprising mistake lol, I guess I will add the proper remap to to both parts, GUI and term usage. Thanks for commenting and pointing that out.

The incomplete commit
https://github.com/rbreaves/kinto/commit/2ffe6390243ac508c646cd2e02a6f3484ee45198

I'll reference this ticket in the next commit.

**Comment by rbreaves** [comment]
Ok, a new update has been merged into master that should address it and I have expanded it to just simple undo any remapping on the right side of the keyboard, if AltGr is back to normal then I feel like it'd throw everything else off so it'd be better just to leave all of it untouched on that side.

If you have a better suggestion then I am all ears. I have also updated the installer to be clearer on what's occurring with it as well - although it should probably say Physical Right Alt key as the virtual Right Alt key should work as well and may lead to the impression that it doesn't work otherwise.
